### PR TITLE
Yearn_v2 connector

### PR DESCRIPTION
**Pull request by TBouder**
**Protocol**: https://yearn.finance
**Docs**: https://docs.yearn.finance/

This connector adds support for interacting with the Yearn Vaults, aka
- calling the `deposit` function to add some funds to a Yearn Vault
- calling the `withdraw` function to redeem the underlying token from the shares

Some basic tests has been added to this connector to deposit & withdraw DAI from/out of [this vault](https://etherscan.io/address/0xdA816459F1AB5631232FE5e97a05BBBb94970c95).

Note: on the withdraw side, an optional argument has been omitted (maxLoss) but could be added on demand.

### CODE REVIEW

**Comment by thrilok209** [comment]
Hey @TBouder! We would also need a resolver. That is used to fetch protocol and connector functions related data.
https://github.com/Instadapp/dsa-resolvers
https://www.notion.so/instadapp/Instadapp-Protocol-Integration-d14f34f1a4aa42138154143948ddb7b4

Eg: Compound resolver- https://github.com/Instadapp/dsa-resolvers/tree/main/contracts/protocols/mainnet/compound

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+        TokenInterface tokenContract = TokenInterface(want);
+
+        _amt = _amt == uint(-1) ? tokenContract.balanceOf(address(this)) : _amt;
+        approve(tokenContract, vault, _amt);
+
+        uint256 _shares = yearn.deposit(_amt, address(this));
+        setUint(setId, _shares);
+
+        _eventName = "LogDeposit(address, uint256, uint256, uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]:
```suggestion
        _eventName = "LogDeposit(address,uint256,uint256,uint256)";
```

@TBouder In _eventName, we can't have space between data types. It changes the encoding of event.

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        uint _amt = getUint(getId, amt);
+
+        YearnV2Interface yearn = YearnV2Interface(vault);
+        TokenInterface tokenContract = TokenInterface(vault);
+
+        _amt = _amt == uint(-1) ? tokenContract.balanceOf(address(this)) : _amt;
+        uint256 _wantRedeemed = yearn.withdraw(_amt, address(this));
+        setUint(setId, _wantRedeemed);
+
+        _eventName = "LogWithdraw(address, uint256, uint256, uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]: @TBouder Same here: 
```suggestion
        _eventName = "LogWithdraw(address,uint256,uint256,uint256)";
```

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        _eventParam = abi.encode(vault, _amt, getId, setId);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]: @TBouder I think we can emit both amount and shares in the deposit and withdraw event.

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        _eventName = "LogDeposit(address, uint256, uint256, uint256)";
+        _eventParam = abi.encode(vault, _amt, getId, setId);
+    }
+
+    /**
+     * @dev Withdraw shares from the vault.
+     * @notice This will withdraw the share from a specific Yearn Vault.
+     * @param vault The address of the vault to withdraw shares from.
+     * @param amt The amount of shares to withdraw.
+     * @param getId ID to retrieve amtA.
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]:
```suggestion
     * @param getId ID to retrieve amt.
```

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import { Events } from "./events.sol";
+import { YearnV2Interface } from "./interface.sol";
+
+abstract contract YearnResolver is Events, Basic {
+    /**
+     * @dev Deposit funds in the vault, issuing shares to recipient.
+     * @notice This will deposit funds to a specific Yearn Vault.
+     * @param vault The address of the vault to deposit funds into.
+     * @param amt The amount of tokens to deposit.
+     * @param getId ID to retrieve amtA.
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]:
```suggestion
     * @param getId ID to retrieve amt.
```

**Comment by TBouder** [comment]
@thrilok209 I updated the elements, thank you for the review

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        uint256 getId,
+        uint256 setId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        uint _amt = getUint(getId, amt);
+
+        YearnV2Interface yearn = YearnV2Interface(vault);
+
+        address want = yearn.token();
+
+        TokenInterface tokenContract = TokenInterface(want);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol:34]: @TBouder Last minor change would be adding logics to convert ETH into WETH if want token is WETH. Eg: https://github.com/Instadapp/dsa-connectors/blob/main/contracts/mainnet/connectors/aave/v2/main.sol#L33-L45

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        uint256 getId,
+        uint256 setId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        uint _amt = getUint(getId, amt);
+
+        YearnV2Interface yearn = YearnV2Interface(vault);
+        TokenInterface tokenContract = TokenInterface(vault);
+
+        _amt = _amt == uint(-1) ? tokenContract.balanceOf(address(this)) : _amt;
+        uint256 _wantRedeemed = yearn.withdraw(_amt, address(this));
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]: @TBouder Same over here: Convert WETH into ETH if want token is WETH. Eg: https://github.com/Instadapp/dsa-connectors/blob/cf504594be9a99e628dd1ba6a7ae8cae46830e9f/contracts/mainnet/connectors/aave/v2/main.sol#L87

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+     * @dev Deposit funds in the vault, issuing shares to recipient.
+     * @notice This will deposit funds to a specific Yearn Vault.
+     * @param token The address of the token to deposit.
+     * @param vault The address of the vault to deposit funds into.
+     * @param amt The amount of tokens to deposit.
+     * @param getId ID to retrieve amt.
+     * @param setId ID stores the amount of shares received.
+    */
+    function deposit(
+        address token,
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]: @TBouder Just curious to know, why the `token` param is needed? We can simply apply the logics using `want`?

File: contracts/mainnet/connectors/yearn_v2/main.sol [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        uint256 amt,
+        uint256 getId,
+        uint256 setId
+    ) external payable returns (string memory _eventName, bytes memory _eventParam) {
+        uint _amt = getUint(getId, amt);
+
+        YearnV2Interface yearn = YearnV2Interface(vault);
+
+        address want = yearn.token();
+        bool isEth = want == ethAddr;
```
Comment (thrilok209) [inline contracts/mainnet/connectors/yearn_v2/main.sol]:
```suggestion
        bool isEth = want == wethAddr;
```
@TBouder I think it should be `wethAddr`?

**Comment by TBouder** [comment]
Connector deployed

ConnectorName: "YEARN-VAULT-A"
ConnectorAddress: [0x598522C536612550bE09F01De201a6634dafA307](https://etherscan.io/address/0x598522c536612550be09f01de201a6634dafa307) !

**Comment by thrilok209** [comment]
Connector whitelisted: https://etherscan.io/tx/0x6337b67cb65e253836f7bc30ba2dc40c5d0c6553bf576c135df93d9b1fed3a44

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**